### PR TITLE
Introduce schema for `pjson` format

### LINKED ISSUE #1581: Modify `pjson` format
Currently, `pjson` stores the value's type together with the value itself:

```json
{
    "$k": ["doc", "array", "d", "s", "bin", "oid", "bool", "t", "null", "regexp", "i32", "ts", "i64"],
    "doc": {
        "$k": ["s"],
        "s": "foo"
    },
    "array": [1, 2],
    "d": {"$f": 42.13},
    "s": "bar",
    "bin": {"$b": "YmF6", "s": 0},
    "oid": {"$o": "010101010101010101010101"},
    "bool": true,
    "t": {"$d": -62167219200000},
    "null": null,
    "regex": {"$r": "hoffman", "o": "i"},
    "i32": 42,
    "ts": {"$t": "1"},
    "i64": {"$l": "9223372036854775807"}
}
```

That limits our ability to pushdown queries and index that data effectively. For example, in queries, MongoDB often does not differentiate between numeric types, but that format does.

Let's extract the whole document's schema into its own field and store values as simpler JSON:

```json
{
    "$s": {
      "$k": ["doc", "array", "d", "s", "bin", "oid", "bool", "t", "null", "regexp", "i32", "ts", "i64"],
      "p": { 
        "doc": {
            "t": "object",
            "$s": {
                "$k": ["s"],
                "s": {"t": "string"}
            }
        },
        "array": {
            "t": "array",
            "i": ["int", "int"]
        },
        "d": {"t": "double"},
        "s": {"t": "string"},
        "bin": {"t": "binData", "s": 0},
        "oid": {"t": "objectId"},
        "bool": {"t": "bool"},
        "t": {"t": "date"},
        "null": {"t": "null"},
        "regexp": {"t": "regex","o": "i"},
        "i32": {"t": "int"},
        "ts": {"t": "timestamp"},
        "i64": {"t": "long"}
      }
    },
    "doc": {
        "s": "foo"
    },
    "array": [1, 2],
    "d": 42.13,
    "s": "bar",
    "bin": "YmF6",
    "oid": "010101010101010101010101",
    "bool": true,
    "t": -62167219200000,
    "null": null,
    "regex": "hoffman",
    "i32": 42,
    "ts": 1,
    "i64": 9223372036854775807
}
```

Notes:

- [x] subdocuments don't have their own schema; it is a part of the main document schema;
- [x] array's schema is decided
- [x] value types currently match MongoDB type aliases, but that might change in the future; we should not try to reuse those constants;
- [x] scalar values are stored as JSON scalars; their options are a part of the schema;
- [x] int64 values should be stored as numbers; PostgreSQL could handle them.
- [x] 0nly `pjson` and `pgdb` (for query pushdown) packages should be updated. If something else should be changed, that's a bug that should be either fixed or a ticket should be created.
- [x] `fjson` package should not be updated. It is not related to `pjson` anymore.

### PR DESCRIPTION
**Pull request by rumyantseva**
# Description

Closes #1581.

This PR is a part of #1620. I moved schema to this separate PR to simplify code review.

**Attention!** This PR changes the format proposed in the original issue. Please review it and agree/disagree. If the new format is ok, I'll modify the description of the original issue.

The idea is quite similar to what we have for `tjson`.

#### Original format
```json
{
        "$k": ["doc", "array"],
        "doc": {
            "t": "object",
            "$s": {
                "$k": ["s"],
                "s": {"t": "string"}
            }
        },
        "array": {
            "t": "array",
            "TODO": "FIXME"
        }
}
```

#### Proposed format
Add `p` field (same level as `$k`) to store properties:
```json
{
        "$k": ["doc", "array"],
        "p": {
               "doc": {
                   "t": "object",
                   "$s": {
                       "$k": ["s"],
                       "s": {"t": "string"}
                   }
              },
              "array": {
                    "t": "array",
                    "i": [{"t": "bool"}, {"t": "date"}, {"t": "regex", "o": "i"}]
             }
        }
}
```

See also `test_schema.go` for more examples.

#### Reasoning

The proposed format doesn't require special handling, a simple `json.Marshal` / `json.Unmarshal` can be used to parse the schema. 

## Readiness checklist

* [x] I added tests for new functionality or bugfixes.
* [ ] I ran `task all`, and it passed.
* [ ] I added/updated comments for both exported and unexported top-level declarations (functions, types, etc).
* [ ] I checked comments rendering with `task godocs`.
* [x] I ensured that the title is good enough for the changelog.
* [x] (for maintainers only) I set Reviewers ([`@FerretDB/core`](https://github.com/orgs/FerretDB/teams/core)), Assignee, Labels, Project and project's Sprint fields.
* [x] I marked all done items in this checklist.

### CODE REVIEW

**Review by w84thesun — APPROVED** [review]
Nice! It was quite easy to read and understand how it works.

File: internal/handlers/pg/pjson/schema_test.go [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+										"bar": nullSchema,
+										"baz": longSchema,
+										"arr": {
+											Type: elemTypeArray,
+											Items: []*elem{
+												intSchema,
+												timestampSchema,
+											},
+										},
+									},
```
Comment (noisersup) [inline internal/handlers/pg/pjson/schema_test.go:57]: woah :smile:

**Review by noisersup — APPROVED** [review]
I like this approach! LGTM!

File: internal/handlers/pg/pjson/schema.go [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+import (
+	"bytes"
+	"encoding/json"
+
+	"github.com/FerretDB/FerretDB/internal/util/lazyerrors"
+)
+
+// schema describes document/object schema needed to unmarshal pjson document.
+type schema struct {
+	Properties map[string]*elem `json:"p"`  // document's properties
```
Comment (chilagrow) [inline internal/handlers/pg/pjson/schema.go:26]: I see that you can't use simple JSON marshal/unmarshal for this case. I want to check if this applies to values of the JSON as well. For example, the values which follows `$s` schema, which will be the next PR. After `$s` we see keys `doc`, `array`, `d`, `s` and `bin`, they are at the same level as `$s`. Can we marshal/unmarshal simply there?

```
   {
    "$s": {
        "$k": ["doc", "array", "d", "s", "bin", "oid", "bool", "t", "null", "regexp", "i32", "ts", "i64"],
        "doc": {
            "t": "object",
            "$s": {
                "$k": ["s"],
                "s": {"t": "string"}
            }
        },
        "array": {
            "t": "array",
            "TODO": "FIXME"
        },
        "d": {"t": "double"},
        "s": {"t": "string"},
        "bin": {"t": "binData", "s": 0},
        "oid": {"t": "objectId"},
        "bool": {"t": "bool"},
        "t": {"t": "date"},
        "null": {"t": "null"},
        "regexp": {"t": "regex","o": "i"},
        "i32": {"t": "int"},
        "ts": {"t": "timestamp"},
        "i64": {"t": "long"}
    },
// here and below
    "doc": {
        "s": "foo"
    },
    "array": [1, 2],
    "d": 42.13,
    "s": "bar",
    "bin": "YmF6",
    "oid": "010101010101010101010101",
    "bool": true,
    "t": -62167219200000,
    "null": null,
    "regex": "hoffman",
    "i32": 42,
    "ts": 1,
    "i64": 9223372036854775807
}
```

**Review by chilagrow — COMMENTED** [review]
Added comments about the new schema change :hugs: Let me know what you think

File: internal/handlers/pg/pjson/schema.go (hunk 2 above)
Comment (rumyantseva) [reply internal/handlers/pg/pjson/schema.go:26]: The problem is not about having or not having `$s` on the same level as values. The problem is having object->array->object->array chains in the schema itself. I tried this approach, which was too complicated (I spent two days and couldn't implement it correctly).

Comment (chilagrow) [reply internal/handlers/pg/pjson/schema.go:26]: Got it, I misunderstood what was complicated. Now got it :pray:

File: internal/handlers/pg/pjson/schema_test.go [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+									"p": {"foo": {"t": "array"}}, "$k": ["foo"]
+								}}
+							]},
+							"empty-arr": {"t": "array"}				
+						},
+						"$k": ["arr", "empty-arr"]
+					}}
+				},
+				"$k": ["obj"]
+			}`,
```
[the remark this answers is not in the document]
Comment (rumyantseva) [reply internal/handlers/pg/pjson/schema_test.go:146]: I'm not sure what exactly the question is about, but maybe this piece of code answers it: https://github.com/FerretDB/FerretDB/pull/1620/files#diff-9e017c02a87be1aa2d092e56dac3c76cb74acddfb2041784bd8e3024c42f38c5R171-R182. So, I don't want to have such endless `json.RawMessage` and then parse `$k` and the rest of such objects :) This is why there are `$k` and `p` on the same level.

[the remark this answers is not in the document]
Comment (chilagrow) [reply internal/handlers/pg/pjson/schema_test.go:146]: Just wanted to make sure that the change was only schema not the value.

[the remark this answers is not in the document]
Comment (chilagrow) [reply internal/handlers/pg/pjson/schema_test.go:146]: Which it is this case.

**Review by chilagrow — APPROVED** [review]

File: internal/handlers/pg/pjson/schema_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (rumyantseva) [reply internal/handlers/pg/pjson/schema_test.go:146]: Well, we do quite a lot of changes in values too, just not in this PR.

[the remark this answers is not in the document]
Comment (rumyantseva) [reply internal/handlers/pg/pjson/schema_test.go:146]: Also, this PR only offers new methods (and tests for them), so we don't break the way how schema/values are stored at the moment.

[the remark this answers is not in the document]
Comment (chilagrow) [reply internal/handlers/pg/pjson/schema_test.go:146]: Yep we wouldn't want to break what we have currently :+1:

**Pull request merged**